src/main/java/org/springframework/samples/petclinic/rest/importcsv/ImportCSV.java

```diff
@@ -26,114 +26,123 @@
-            }
+        return new ResponseEntity<List<Pet>>(pets, HttpStatus.OK);
+    }
 
-            field = "";
-            while (i < csv.length() && csv.charAt(i) != ';') {
-                field += csv.charAt(i++);
-            }
-            i++;
-
-            if (pet != null) {
-                ArrayList<PetType> ts = (ArrayList<PetType>) clinicService.findPetTypes();
-                for (int j = 0; j < ts.size(); j++) {
-                    if (ts.get(j).getName().toLowerCase().equals(field)) {
-                        pet.setType(ts.get(j));
-                        break;
-                    }
-                }
-            }
+    /**
+     * Custom Exception which can arise during csv string parsing, error message
+     * contains where the exception was raised
+     */
+    private class IncorrectFormatException extends Exception {
+        public IncorrectFormatException(String errorMessage) {
+            super(errorMessage);
+        }
+    }
 
-            field = "";
-            while (i < csv.length() && (csv.charAt(i) != ';' && csv.charAt(i) != '\n')) {
-                field += csv.charAt(i++);
+    /**
+     * Method which receives a string from the csv and pars it to a list of pets
+     * which is then returns. Exceptions are thrown upon the error field
+     *
+     * @param csv the csv string from the user
+     * @return pets list of pets
+     * @throws IncorrectFormatException the user input is malformatted
+     */
+    private List<Pet> ParseFields(String csv) throws IncorrectFormatException {
+        int i = 0;
+        List<Pet> pets = new LinkedList<Pet>();
+        Pet pet;
+        pet = new Pet();
+
+        String[] arrayOfCsv = csv.split(";", 0);
+        pet.setName(arrayOfCsv[0]);
+
+        try {
+            pet.setBirthDate((new SimpleDateFormat("yyyy-MM-dd")).parse(arrayOfCsv[1]));
+        } catch (ParseException e) {
+            throw new IncorrectFormatException("Date not valid");
+        }
+
+        ArrayList<PetType> petTypes = (ArrayList<PetType>) clinicService.findPetTypes();
+        for (PetType petType : petTypes) {
+            if (petType.getName().toLowerCase().equals(arrayOfCsv[2])) {
+                pet.setType(petType);
+                break;
             }
-
-            if (pet != null) {
-                String owner = field;
-                List<Owner> matchingOwners = clinicService.findAllOwners()
-                    .stream()
-                    .filter(o -> o.getLastName().equals(owner))
-                    .collect(Collectors.toList());
-
-                if (matchingOwners.size() == 0) {
-                    HttpHeaders headers = new HttpHeaders();
-                    headers.add("errors", "Owner not found");
-                    return new ResponseEntity<List<Pet>>(headers, HttpStatus.BAD_REQUEST);
+        }
+
+        String owner = arrayOfCsv[3];
+        List<Owner> matchingOwners = clinicService.findAllOwners()
```

Comment:
try to let the database to the filtering, because you have to load all data (maybe over network) and then filter it here ... the database is highly optimized to filter such things and only return your desired result

Reply:
then you also would have only one object and not a list with (hopefully) one item.